### DISCUSSION TITLE
Symmetric decryption slower than with private key

### DISCUSSION THREAD

**Discussion by fi4sk0** [openpgpjs/openpgpjs#1481] [Q&A] [answered]
Hey everybody, 

I modified `test/benchmarks/time.js` to also include a test for symmetric decryption: First, I created an armored symmetrically encrypted message in `getTestData`:

```javascript
      const armoredSymmetricEncryptedMessage = await openpgp.encrypt({ message: plaintextMessage, passwords: 'super secret!!' });
```

and then added the following case to the test suite:
```javascript
  suite.add('openpgp.decrypt (symmetric)', wrapAsync(async () => {
    const message = await openpgp.readMessage({ armoredMessage: armoredSymmetricEncryptedMessage });
    await openpgp.decrypt({ message, passwords: 'super secret!!' });
  }));
```

Running the benchmarks yields the following results:
```
openpgp.readKey x 3,981 ops/sec ±4.40% (69 runs sampled)
openpgp.readMessage x 12,714 ops/sec ±1.83% (76 runs sampled)
openpgp.generateKey x 52.26 ops/sec ±4.14% (62 runs sampled)
openpgp.encrypt x 479 ops/sec ±0.82% (81 runs sampled)
openpgp.sign x 214 ops/sec ±4.32% (67 runs sampled)
--> openpgp.decrypt x 726 ops/sec ±1.27% (80 runs sampled)
--> openpgp.decrypt (symmetric) x 23.28 ops/sec ±0.88% (58 runs sampled)
openpgp.verify x 113 ops/sec ±3.63% (62 runs sampled)
```

It seems like there is factor ~30 between symmetric and asymmetric decrypt, where asymmetric is faster. I was under the impression that symmetric decryption should be faster than using a public/private key. Am I doing something wrong? Have I misunderstood 
some basic principles?

I'm happy for any pointers!

**Comment by twiss** [comment, marked as answer]
Hello :wave: Password-based encryption has a step where the key is derived from the password, which is intentionally slow, in order to make it harder for attackers to guess the password via a brute-force attack. With asymmetric encryption, this concern doesn't apply as the entropy of the key itself should already be sufficient. If you're very confident that the password you're using has sufficient entropy as well (e.g. it's a randomly generated string of sufficient length, and not something provided by a user) then you can lower `openpgp.config.s2kIterationCountByte` to lower this protection and increase the performance.

**Reply by fi4sk0** [reply]
Thanks @twiss, this solved my problem! 

Unfortunately, the answer begs the question of how to generate the password (length? eg. base64 encoded random bytes?) and what s2kIterationCountByte is still acceptable. Can you further elaborate?

This is not a pressing matter. My application is still in its infancy and these settings I can easily adjust as I learn more about the topic, so a good read is also very welcome. I already watched and enjoyed Laurens Van Houtvens crypto 101 talk on pycon 2013 and I started reading the book on https://www.crypto101.io but I don't dare to choose any settings where I don't fully understand the implications.

**Reply by twiss** [reply]
The "normal" use case for the password is a user-provided passphrase. However, when randomly generating a password, you would probably want the same amount of entropy as the resulting key, so 32 bytes for AES-256, for example. So you could generate 32 random bytes and then base64-encode them, yes.

Then, if you do so, setting `s2kIterationCountByte` to 0 is acceptable, as then the password has the same amount of entropy as the resulting key, so this protection is not needed. Of course, if the password has less entropy or is user-provided, then setting it to 0 is not secure.

You can also call `openpgp.encrypt({ message, passwords: [randomPassword], config: { s2kIterationCountByte: 0 } })`, to make sure that this config only applies when you pass a random password, and not for other uses of OpenPGP.js.

**Reply by twiss** [reply]
(The keyword to learn more about this is "key derivation function", e.g. https://en.wikipedia.org/wiki/Key_derivation_function)

**Reply by steida** [reply]
@fi4sk0 I am using https://github.com/bitcoinjs/bip39 mnemonic. 12 words have good UX.

@twiss I plan OpenPGP public-private encryption but for user-only data, I am using symmetric encryption with bip39 mnemonic. For such a use case, browsers only, does it make sense to use OpenPGP? Isn't native subtle crypto good enough? How OpenPGP symmetric encryption with s2kIterationCountByte == 0 is different from native subtle crypto? Thank you.

**Reply by twiss** [reply]
12 words will give you 16 bytes / 128 bits of security, similar to the security of AES-128.

If you don't need interoperability with the existing OpenPGP (email) ecosystem, you don't necessarily need OpenPGP.js, no. I would only recommend using subtle crypto if you know what you're doing (hence the name "subtle"), but it's possible to use it securely, yeah.

**Reply by steida** [reply]
@twiss Thank you. Btw, are you watching Project Everest? Will it be used in OpenPGP?

**Reply by twiss** [reply]
Not very likely, unless it's via wasm - but the ideal way for us to end up using it is via the Web Crypto API. I believe Firefox is already using HACL* for some cryptographic algorithms, in which case we'd be using them indirectly if they're exposed in Web Crypto (which unfortunately isn't always the case - e.g. see https://github.com/w3c/webcrypto/issues/196).
